### PR TITLE
Fix types no longer return callbacks

### PR DESCRIPTION
**Pull request by flamerohr**
This opens up the option to upgrade to `graph-php@0.8`

Fixes https://github.com/silverstripe/silverstripe-graphql/issues/47

Dependencies https://github.com/silverstripe/silverstripe-asset-admin/pull/474

### CODE REVIEW

**Comment by tractorcow** [comment]
Wow good job. I thought this work would be a lot harder than it was.

I suggest either @unclecheese or @chillu to review this given their proximity to this module.

**Comment by chillu** [comment]
Yep, as you expected @tractorcow, it's not that easy. I've amended Chris' PR with the other uses where `type()` returns a function rather than a value, and you can see all kinds of unit tests failing.

**Comment by unclecheese** [comment]
Yeah, this change is going to completely hose all the scaffolding stuff. I've been dreading this upgrade.

**Comment by flamerohr** [comment]
Yeah, I didn't think this was the complete work... I was waiting for travis to run unit tests, as they seem to run even slower on my own computer.
I'm still trying out some other things like adding callback wrapper to the `'fields'` property in `ObjectType`.

What I've done so far is only tested with asset-admin currently, and I'm building on top of this as I go along :D Quite new to all these layers of building types and scaffolding in PHP, so it's taking me a bit of time to get the theory into my head.

File: src/Scaffolding/Scaffolders/MutationScaffolder.php [hunk 1]
```diff
@@ -29,6 +29,11 @@ public function addToManager(Manager $manager)
      */
     public function scaffold(Manager $manager)
     {
+        $name = $this->typeName;
+        if (!$manager->hasType($name) && method_exists($this, 'generateInputType')) {
+            $manager->addType($this->generateInputType(), $name);
```
Comment (tractorcow) [inline src/Scaffolding/Scaffolders/MutationScaffolder.php]: I think we need to move this to an interface, and then use `$this instanceof TheInterface` to promote good type cohesion. On the interface you can explain what this type does. :)

Comment (unclecheese) [inline src/Scaffolding/Scaffolders/MutationScaffolder.php]: Interfaces can't use non-public methods. :(

**Pull request merged**